File: CS/Dennis.Search.Win/SearchObjectViewController.cs
using System;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.DC;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Utils;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.SystemModule;

namespace Dennis.Search.Win {
    public class SearchObjectViewController : ViewController {
        private const string ActiveKeySearchObject = "ISearchObject";
        private SingleChoiceAction scaSearchObjectActionCore;
        public const string SearchResultsCollectionName = "SearchResults";
        public SearchObjectViewController() {
            scaSearchObjectActionCore = new SingleChoiceAction(this, "SearchObject", PredefinedCategory.Search);
            scaSearchObjectActionCore.Execute += saSearchObjectActionCore_Execute;
            scaSearchObjectActionCore.Caption = Properties.Resources.SearchObjectActionText;
            scaSearchObjectActionCore.ImageName = "Action_Search";
            scaSearchObjectActionCore.ItemType = SingleChoiceActionItemType.ItemIsOperation;
        }
        public SingleChoiceAction SearchObjectAction {
            get { return scaSearchObjectActionCore; }
        }
        private void saSearchObjectActionCore_Execute(object sender, SingleChoiceActionExecuteEventArgs args) {
            SearchObject(args);
        }
        protected virtual void SearchObject(SingleChoiceActionExecuteEventArgs args) {
            Type type = (Type)args.SelectedChoiceActionItem.Data;
            IObjectSpace os = Application.CreateObjectSpace(type);
            object obj = os.CreateObject(type);
            DetailView dv = Application.CreateDetailView(os, obj);
            args.ShowViewParameters.CreatedView = dv;
        }
        protected override void OnActivated() {
            base.OnActivated();
            InitSearchObjectActionItems();
            if(typeof(ISearchObject).IsAssignableFrom(View.ObjectTypeInfo.Type)) {
                DetailView dv = View as DetailView;
                if(dv != null) {
                    foreach(ListPropertyEditor editor in dv.GetItems<ListPropertyEditor>())
                        if(editor.PropertyName == SearchResultsCollectionName)
                            editor.ControlCreated += new EventHandler<EventArgs>(editor_ControlCreated);
                    RefreshController refreshController = Frame.GetController<RefreshController>();
                    if(refreshController != null) {
                        refreshController.RefreshAction.Active[ActiveKeySearchObject] = false;
                    }
                    RecordsNavigationController recordsNavigationController = Frame.GetController<RecordsNavigationController>();
                    if(recordsNavigationController != null) {
                        recordsNavigationController.PreviousObjectAction.Active[ActiveKeySearchObject] = false;
                        recordsNavigationController.NextObjectAction.Active[ActiveKeySearchObject] = false;
                    }
                    ModificationsController modificationsController = Frame.GetController<ModificationsController>();
                    if(modificationsController != null) {
                        modificationsController.SaveAction.Active[ActiveKeySearchObject] = false;
                        modificationsController.SaveAndCloseAction.Active[ActiveKeySearchObject] = false;
                        modificationsController.CancelAction.Active[ActiveKeySearchObject] = false;
                    }
                    DeleteObjectsViewController deleteObjectsViewController = Frame.GetController<DeleteObjectsViewController>();
                    if(deleteObjectsViewController != null) {
                        deleteObjectsViewController.DeleteAction.Active[ActiveKeySearchObject] = false;
                    }
                    NewObjectViewController newObjectViewController = Frame.GetController<NewObjectViewController>();
                    if(newObjectViewController != null) {
                        newObjectViewController.NewObjectAction.Active[ActiveKeySearchObject] = false;
                    }
                }
                SearchObjectAction.Active[ActiveKeySearchObject] = false;
            }
        }
        private void editor_ControlCreated(object sender, EventArgs e) {
            ListPropertyEditor editor = (ListPropertyEditor)sender;
            editor.ControlCreated -= new EventHandler<EventArgs>(editor_ControlCreated);
            LinkUnlinkController linkUnlinkController = editor.Frame.GetController<LinkUnlinkController>();
            if(linkUnlinkController != null) {
                linkUnlinkController.LinkAction.Active[ActiveKeySearchObject] = false;
                linkUnlinkController.UnlinkAction.Active[ActiveKeySearchObject] = false;
            }
            DeleteObjectsViewController deleteObjectsViewController = editor.Frame.GetController<DeleteObjectsViewController>();
            if(deleteObjectsViewController != null) {
                deleteObjectsViewController.DeleteAction.Active[ActiveKeySearchObject] = false;
            }
            NewObjectViewController newObjectViewController = editor.Frame.GetController<NewObjectViewController>();
            if(newObjectViewController != null) {
                newObjectViewController.NewObjectAction.Active[ActiveKeySearchObject] = false;
            }
        }
        private void InitSearchObjectActionItems() {
            SearchObjectAction.BeginUpdate();
            SearchObjectAction.Items.Clear();
            Type genericType = typeof(SearchObjectBase<>).MakeGenericType(View.ObjectTypeInfo.Type);
            string imageName = View.Model.ImageName;
            foreach (ITypeInfo typeInfo in XafTypesInfo.Instance.PersistentTypes) {
                if (genericType.IsAssignableFrom(typeInfo.Type) && !typeInfo.IsAbstract) {
                    ChoiceActionItem item = new ChoiceActionItem(CaptionHelper.GetClassCaption(typeInfo.FullName), typeInfo.Type);
                    item.ImageName = imageName;
                    SearchObjectAction.Items.Add(item);
                }
            }
            SearchObjectAction.EndUpdate();
        }
    }
}

File: VB/Dennis.Search.Win/SearchObjectViewController.vb
Imports System
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.DC
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Utils
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.SystemModule

Namespace Dennis.Search.Win
    Public Class SearchObjectViewController
        Inherits ViewController

        Private Const ActiveKeySearchObject As String = "ISearchObject"
        Private scaSearchObjectActionCore As SingleChoiceAction
        Public Const SearchResultsCollectionName As String = "SearchResults"
        Public Sub New()
            scaSearchObjectActionCore = New SingleChoiceAction(Me, "SearchObject", PredefinedCategory.Search)
            AddHandler scaSearchObjectActionCore.Execute, AddressOf saSearchObjectActionCore_Execute
            scaSearchObjectActionCore.Caption = My.Resources.SearchObjectActionText
            scaSearchObjectActionCore.ImageName = "Action_Search"
            scaSearchObjectActionCore.ItemType = SingleChoiceActionItemType.ItemIsOperation
        End Sub
        Public ReadOnly Property SearchObjectAction() As SingleChoiceAction
            Get
                Return scaSearchObjectActionCore
            End Get
        End Property
        Private Sub saSearchObjectActionCore_Execute(ByVal sender As Object, ByVal args As SingleChoiceActionExecuteEventArgs)
            SearchObject(args)
        End Sub
        Protected Overridable Sub SearchObject(ByVal args As SingleChoiceActionExecuteEventArgs)
            Dim type As Type = CType(args.SelectedChoiceActionItem.Data, Type)
            Dim os As IObjectSpace = Application.CreateObjectSpace(type)
            Dim obj As Object = os.CreateObject(type)
            Dim dv As DetailView = Application.CreateDetailView(os, obj)
            args.ShowViewParameters.CreatedView = dv
        End Sub
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            InitSearchObjectActionItems()
            If GetType(ISearchObject).IsAssignableFrom(View.ObjectTypeInfo.Type) Then
                Dim dv As DetailView = TryCast(View, DetailView)
                If dv IsNot Nothing Then
                    For Each editor As ListPropertyEditor In dv.GetItems(Of ListPropertyEditor)()
                        If editor.PropertyName = SearchResultsCollectionName Then
                            AddHandler editor.ControlCreated, AddressOf editor_ControlCreated
                        End If
                    Next editor
                    Dim refreshController As RefreshController = Frame.GetController(Of RefreshController)()
                    If refreshController IsNot Nothing Then
                        refreshController.RefreshAction.Active(ActiveKeySearchObject) = False
                    End If
                    Dim recordsNavigationController As RecordsNavigationController = Frame.GetController(Of RecordsNavigationController)()
                    If recordsNavigationController IsNot Nothing Then
                        recordsNavigationController.PreviousObjectAction.Active(ActiveKeySearchObject) = False
                        recordsNavigationController.NextObjectAction.Active(ActiveKeySearchObject) = False
                    End If
                    Dim modificationsController As ModificationsController = Frame.GetController(Of ModificationsController)()
                    If modificationsController IsNot Nothing Then
                        modificationsController.SaveAction.Active(ActiveKeySearchObject) = False
                        modificationsController.SaveAndCloseAction.Active(ActiveKeySearchObject) = False
                        modificationsController.CancelAction.Active(ActiveKeySearchObject) = False
                    End If
                    Dim deleteObjectsViewController As DeleteObjectsViewController = Frame.GetController(Of DeleteObjectsViewController)()
                    If deleteObjectsViewController IsNot Nothing Then
                        deleteObjectsViewController.DeleteAction.Active(ActiveKeySearchObject) = False
                    End If
                    Dim newObjectViewController As NewObjectViewController = Frame.GetController(Of NewObjectViewController)()
                    If newObjectViewController IsNot Nothing Then
                        newObjectViewController.NewObjectAction.Active(ActiveKeySearchObject) = False
                    End If
                End If
                SearchObjectAction.Active(ActiveKeySearchObject) = False
            End If
        End Sub
        Private Sub editor_ControlCreated(ByVal sender As Object, ByVal e As EventArgs)
            Dim editor As ListPropertyEditor = DirectCast(sender, ListPropertyEditor)
            RemoveHandler editor.ControlCreated, AddressOf editor_ControlCreated
            Dim linkUnlinkController As LinkUnlinkController = editor.Frame.GetController(Of LinkUnlinkController)()
            If linkUnlinkController IsNot Nothing Then
                linkUnlinkController.LinkAction.Active(ActiveKeySearchObject) = False
                linkUnlinkController.UnlinkAction.Active(ActiveKeySearchObject) = False
            End If
            Dim deleteObjectsViewController As DeleteObjectsViewController = editor.Frame.GetController(Of DeleteObjectsViewController)()
            If deleteObjectsViewController IsNot Nothing Then
                deleteObjectsViewController.DeleteAction.Active(ActiveKeySearchObject) = False
            End If
            Dim newObjectViewController As NewObjectViewController = editor.Frame.GetController(Of NewObjectViewController)()
            If newObjectViewController IsNot Nothing Then
                newObjectViewController.NewObjectAction.Active(ActiveKeySearchObject) = False
            End If
        End Sub
        Private Sub InitSearchObjectActionItems()
            SearchObjectAction.BeginUpdate()
            SearchObjectAction.Items.Clear()
            Dim genericType As Type = GetType(SearchObjectBase(Of )).MakeGenericType(View.ObjectTypeInfo.Type)
            Dim imageName As String = View.Model.ImageName
            For Each typeInfo As ITypeInfo In XafTypesInfo.Instance.PersistentTypes
                If genericType.IsAssignableFrom(typeInfo.Type) AndAlso Not typeInfo.IsAbstract Then
                    Dim item As New ChoiceActionItem(CaptionHelper.GetClassCaption(typeInfo.FullName), typeInfo.Type)
                    item.ImageName = imageName
                    SearchObjectAction.Items.Add(item)
                End If
            Next typeInfo
            SearchObjectAction.EndUpdate()
        End Sub
    End Class
End Namespace